File: recurpicker/src/main/java/com/maltaisn/recurpicker/RecurrencePickerSettings.java
/*
 * Copyright (c) Nicolas Maltais
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package com.maltaisn.recurpicker;


import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import java.text.DateFormat;

/**
 * As RecurrencePickerView and RecurrencePickerDialog are using the same methods
 * for the settings, this interface improves consistency
 */
public interface RecurrencePickerSettings {

    int MAX_FIELD_VALUE = 999999999;
    int DEFAULT_MAX_END_COUNT = 999;
    int DEFAULT_MAX_FREQUENCY = 99;
    int DEFAULT_MAX_END_DATE = -1;
    int DEFAULT_END_COUNT = 5;
    boolean DEFAULT_END_DATE_USE_PERIOD = true;
    int DEFAULT_END_DATE_INTERVAL = 3;
    boolean DEFAULT_OPTION_LIST_ENABLED = true;
    boolean DEFAULT_CREATOR_ENABLED = true;
    boolean DEFAULT_SHOW_DONE_IN_LIST = false;
    boolean DEFAULT_SHOW_HEADER_IN_LIST = true;
    boolean DEFAULT_SHOW_CANCEL_BTN = false;
    int DEFAULT_ENABLED_PERIODS = 0b1111;
    int DEFAULT_ENABLED_END_TYPES = 0b111;
    Recurrence[] DEFAULT_OPTION_LIST_DEFAULTS = null;
    CharSequence[] DEFAULT_OPTION_LIST_TITLES = null;

    /**
     * Set the date format to use
     * By default, the system default date format is used
     * You should call this method before {@link #setRecurrence(Recurrence, long)}
     * @param endDateFormat Date format to use for the end date
     *                      This date should usually be shorter
     * @param optionListDateFormat Date format to use for the custom item in the option list
     *                         This date should usually be more verbose
     */
    RecurrencePickerSettings setDateFormat(@NonNull DateFormat endDateFormat, @NonNull DateFormat optionListDateFormat);

    /**
     * Call this method to initialize the recurrence picker with a recurrence and a date
     * @param recurrence Recurrence to display, use null for the default "does not repeat"
     * @param startDate Starting date of recurrence to be returned, use 0 for today
     */
    RecurrencePickerSettings setRecurrence(@Nullable Recurrence recurrence, long startDate);

    /**
     * Set the maximum event count for a recurrence
     * By default, the maximum number of events is 999
     * Must be set before showing the creator or the dialog
     * @param max Maximum count, set to -1 to allow any number of events
     *            Even if set to -1, maximum is {@link RecurrencePickerView#MAX_FIELD_VALUE}
     */
    RecurrencePickerSettings setMaxEventCount(int max);

    /**
     * Set the maximum frequency for a period
     * By default, maximum frequency is 99
     * Must be set before showing the creator or the dialog
     * ex: if 10, maximum will be: every 10 days, every 10 weeks, every 10 months...
     * @param max Maximum frequency, set to -1 to allow any value
     *            Even if set to -1, maximum is {@link RecurrencePickerView#MAX_FIELD_VALUE}
     */
    RecurrencePickerSettings setMaxFrequency(int max);

    /**
     * Set the maximum end date for a recurrence
     * By default, there is no maximum end date
     * Must be set before showing the creator or the dialog
     * @param time Time in millis for a date, set to -1 for no maximum
     */
    RecurrencePickerSettings setMaxEndDate(long time);

    /**
     * Change how the end date is set relative to the start date by default
     * By default, default end date is 3 periods after start date
     * Must be set before showing the creator or the dialog
     * @param usePeriod If true, will use the currently selected period as the unit of the interval
     *                  ex: if repeating weekly and interval is 3, default end date will be 3 weeks after start date
     *                  If false, will use days as the unit of the interval
     * @param interval How many periods/days after start date to set default end date
     */
    RecurrencePickerSettings setDefaultEndDate(boolean usePeriod, int interval);

    /**
     * Set the number of events to show by default
     * By default, recurrence ends after 5 events
     * Must be set before showing the creator or the dialog
     * @param count Number of events to show by default
     */
    RecurrencePickerSettings setDefaultEndCount(int count);

    /**
     * Select which modes are enabled: default options list and recurrence creator
     * By default, both modes are enabled, and at least one of them should be
     * @param optionListEnabled whether to enable the default options list
     * @param creatorEnabled whether to enabled the recurrence creator
     */
    RecurrencePickerSettings setEnabledModes(boolean optionListEnabled, boolean creatorEnabled);

    /**
     * Set whether to show the Done button in the first screen, the list of default recurrence options,
     * or to hide it. If shown, listener won't be called when option is clicked, only when button is.
     * By default, done button is not shown in option list.
     * @param show Whether to show it or not
     */
    RecurrencePickerSettings setShowDoneButtonInOptionList(boolean show);

    /**
     * Set whether the header should be displayed in the first screen, the list of default recurrence options,
     * or not. If you wish to also hide the header in the creator mode, you can change its background color.
     * By default, header is shown.
     * @param show Whether to show it or not
     */
    RecurrencePickerSettings setShowHeaderInOptionList(boolean show);

    /**
     * Set whether to show a cancel button or not. Cancel button is placed at the left side of the done button.
     * By default, cancel button is hidden.
     * @param show Whether to show it or not
     */
    RecurrencePickerSettings setShowCancelButton(boolean show);

    /**
     * Set which periods are available to be used. Note that NONE period cannot be disabled
     * You must set at least one period.
     * Must be set before showing the creator or the dialog
     * @param periods bit field of 1 << [Recurrence.DAILY to Recurrence.YEARLY]. Ex:
     *                periods = (1 << Recurrence.DAILY) | (1 << Recurrence.WEEKLY)
     *                Don't forget the parentheses!
     *                A quicker way would be to use binary literal ex: periods = 0b1111 (all periods)
     */
    RecurrencePickerSettings setEnabledPeriods(int periods);

    /**
     * Set which end types are available to be used. Note that END_BY_DATE_OR_COUNT is not implemented
     * You must set at least one type. If that's the case, end type spinner will be disabled.
     * If the only end type set is forever, end type spinner completely disappears.
     * Must be set before showing the creator or the dialog
     * @param types bit field of 1 << [Recurrence.END_NEVER, END_BY_DATE and END_BY_COUNT]. Ex:
     *              periods = (1 << Recurrence.END_BY_DATE) | (1 << Recurrence.END_BY_COUNT)
     *              Don't forget the parentheses!
     *              A quicker way would be to use binary literal ex: types = 0b111 (all types)
     */
    RecurrencePickerSettings setEnabledEndTypes(int types);

    /**
     * Change defaults in option list. Note that "Does not repeat" will always be shown.
     * By default, a daily, weekly, monthly and yearly recurrences with frequency of 1 are used
     * MUST be set before setting the recurrence if using the view by itself
     * @param defaults Array of recurrence to use as defaults, leave null to use default defaults (exactly)
     *                 or leave null to only change the titles
     *                 You can set any start date for these recurrences because it gets changed
     * @param titles Array of titles to use for each recurrence, leave null to use
     *               {@link RecurrenceFormat#format(Recurrence)} instead. You can also leave
     *               specific items in array null to only format those.
     */
    RecurrencePickerSettings setOptionListDefaults(@Nullable Recurrence[] defaults, @Nullable CharSequence[] titles);

}

File: lib/src/main/kotlin/com/maltaisn/recurpicker/RecurrencePickerSettings.kt
/*
 * Copyright 2019 Nicolas Maltais
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

@file:Suppress("MagicNumber")

package com.maltaisn.recurpicker

import android.os.Bundle
import android.os.Parcel
import android.os.Parcelable
import android.util.Log
import androidx.core.os.BundleCompat
import com.maltaisn.recurpicker.Recurrence.Period
import com.maltaisn.recurpicker.RecurrencePickerSettings.Builder
import com.maltaisn.recurpicker.format.RecurrenceFormatter
import java.text.DateFormat
import java.text.SimpleDateFormat
import java.util.Locale

/**
 * A class for configuring the behavior of the recurrence list and picker dialogs.
 * The class is immutable, use the [Builder] to create it.
 */
public class RecurrencePickerSettings private constructor(
    /**
     * The recurrence formatter used to format the recurrences shown in the recurrence list dialog.
     * The date format is also used to format the end dates in the recurrence picker.
     */
    public val formatter: RecurrenceFormatter,

    /**
     * List of recurrence presets displayed in the recurrence list dialog.
     * A `null` recurrence will be replaced by a "Custom..." item to customize recurrence.
     * Hence, to disable the custom recurrence picker, set no `null` items.
     * The default is a list of recurrences equivalent to "Does not repeat", "Daily",
     * "Weekly", "Monthly", "Yearly", null ("Custom...").
     */
    public val presets: List<Recurrence?>,

    /**
     * The recurrence shown by default in the recurrence picker.
     * This is shown when no recurrence or a "Does not repeat" recurrence is selected.
     */
    public val defaultPickerRecurrence: Recurrence,

    /** The maximum frequency that can be entered in the picker. Must be at least 1. */
    public val maxFrequency: Int,

    /** The maximum end count that can be entered in the picker. Must be at least 1. */
    public val maxEndCount: Int
) : Parcelable {

    public class Builder {

        /** @see RecurrencePickerSettings.formatter */
        @set:JvmSynthetic
        public var formatter: RecurrenceFormatter = RecurrenceFormatter(DateFormat.getDateInstance())

        /** @see RecurrencePickerSettings.presets */
        @set:JvmSynthetic
        public var presets: List<Recurrence?> = mutableListOf(
            Recurrence.DOES_NOT_REPEAT,
            Recurrence(Period.DAILY),
            Recurrence(Period.WEEKLY),
            Recurrence(Period.MONTHLY),
            Recurrence(Period.YEARLY),
            null)

        /** @see RecurrencePickerSettings.defaultPickerRecurrence */
        @set:JvmSynthetic
        public var defaultPickerRecurrence: Recurrence = Recurrence(Period.DAILY)

        /** @see RecurrencePickerSettings.maxFrequency */
        @set:JvmSynthetic
        public var maxFrequency: Int = 99
            set(value) {
                require(value >= 1) { "Max frequency must be at least 1." }
                field = value
            }

        /** @see RecurrencePickerSettings.maxEndCount */
        @set:JvmSynthetic
        public var maxEndCount: Int = 999
            set(value) {
                require(value >= 1) { "Max end count must be at least 1." }
                field = value
            }

        public fun setFormatter(formatter: RecurrenceFormatter): Builder = apply { this.formatter = formatter }
        public fun setPresets(presets: List<Recurrence?>): Builder = apply { this.presets = presets }
        public fun setDefaultPickerRecurrence(recurrence: Recurrence): Builder =
            apply { defaultPickerRecurrence = recurrence }

        public fun setMaxFrequency(frequency: Int): Builder = apply { maxFrequency = frequency }
        public fun setMaxEndCount(count: Int): Builder = apply { maxEndCount = count }

        public fun build(): RecurrencePickerSettings = RecurrencePickerSettings(formatter, presets,
            defaultPickerRecurrence, maxFrequency, maxEndCount)
    }

    // Parcelable stuff
    override fun writeToParcel(parcel: Parcel, flags: Int) {
        val bundle = Bundle()
        bundle.putRecurrenceFormatter(formatter)
        bundle.putParcelableArrayList("presets", ArrayList(presets))
        bundle.putParcelable("defaultPickerRecurrence", defaultPickerRecurrence)
        bundle.putInt("maxFrequency", maxFrequency)
        bundle.putInt("maxEndCount", maxEndCount)
        parcel.writeBundle(bundle)
    }

    override fun describeContents(): Int = 0

    public companion object {
        @JvmField
        public val CREATOR: Parcelable.Creator<RecurrencePickerSettings> =
            object : Parcelable.Creator<RecurrencePickerSettings> {
                override fun createFromParcel(parcel: Parcel) = RecurrencePickerSettings {
                    val bundle = parcel.readBundle(RecurrencePickerSettings::class.java.classLoader)!!
                    formatter = bundle.getRecurrenceFormatter()
                    presets = BundleCompat.getParcelableArrayList(bundle, "presets", Recurrence::class.java)!!
                    defaultPickerRecurrence = BundleCompat.getParcelable(
                        bundle, "defaultPickerRecurrence", Recurrence::class.java)!!
                    maxFrequency = bundle.getInt("maxFrequency")
                    maxEndCount = bundle.getInt("maxEndCount")
                }

                override fun newArray(size: Int) = arrayOfNulls<RecurrencePickerSettings>(size)
            }

        /**
         * Utility function to create settings using constructor-like syntax.
         */
        public inline operator fun invoke(init: Builder.() -> Unit = {}): RecurrencePickerSettings =
            Builder().apply(init).build()

        private fun Bundle.getRecurrenceFormatter(): RecurrenceFormatter {
            val dateFormat = try {
                this.getSerializableCompat<DateFormat>("dateFormat")!!
            } catch (e: Exception) {
                // Very rarely and on API >= 28, Bundle will fail to get serialized DateFormat.
                // This issue is related to: https://stackoverflow.com/a/54155356/5288316.
                // Luckily, DateFormat is most often a SimpleDateFormat, which can be saved
                // using a string. Note that the naming of the key is important here. I had
                // initially named the key "dateFormatPattern", which led getString to fail
                // like getSerializable did!
                if (this.containsKey("dfPattern")) {
                    SimpleDateFormat(this.getString("dfPattern", ""), Locale.getDefault())
                } else {
                    // Ok, date format was lost for good. Use default.
                    Log.e(TAG, "RecurrencePickerSettings formatter's date format lost during unparcelization.")
                    DateFormat.getDateInstance()
                }
            }
            return RecurrenceFormatter(dateFormat)
        }

        private fun Bundle.putRecurrenceFormatter(formatter: RecurrenceFormatter) {
            val dateFormat = formatter.dateFormat
            this.putSerializable("dateFormat", dateFormat)
            if (dateFormat is SimpleDateFormat) {
                this.putString("dfPattern", dateFormat.toPattern())
            }
        }

        private val TAG = RecurrencePickerSettings::class.java.simpleName
    }
}
